arangod/Replication2/ReplicatedLog/Components/StorageManager.cpp

```diff
@@ -0,0 +1,358 @@
+                                                IStorageEngineMethods&>);
+    futures::Future<Result> run(
+        IStorageEngineMethods& methods) noexcept override {
+      return F::operator()(methods);
+    }
+  };
+
+  return scheduleOperation(
+      std::move(guard), std::move(result), std::move(mapResult),
+      std::make_unique<LambdaOperation>(std::forward<F>(fn)));
+}
+
+void StorageManager::triggerQueueWorker(GuardType guard) noexcept {
+  auto const worker = [](GuardType guard,
+                         std::shared_ptr<StorageManager> self) noexcept
+      -> futures::Future<futures::Unit> {
+    LOG_CTX("6efe9", TRACE, self->loggerContext)
+        << "starting new storage worker";
+    while (true) {
+      if (guard->queue.empty()) {
+        LOG_CTX("882a0", TRACE, self->loggerContext)
+            << "stopping storage worker";
+        guard->workerActive = false;
+        break;
+      }
+
+      auto req = std::move(guard->queue.front());
+      guard->queue.pop_front();
+      if (not guard->methods) {
+        guard.unlock();
+        LOG_CTX("4f5e3", DEBUG, self->loggerContext)
+            << "aborting storage operation "
+            << " because log core gone";
+        req.promise.setValue(
+            TRI_ERROR_REPLICATION_REPLICATED_LOG_PARTICIPANT_GONE);
+        guard = self->guardedData.getLockedGuard();
+        continue;
+      }
+      LOG_CTX("e0a6d", TRACE, self->loggerContext)
+          << "executing storage operation";
+      auto f = req.operation->run(*guard->methods);
+      guard.unlock();
+      Result result = co_await asResult(std::move(f));
+      // lock guard to make effects visible
+      if (result.ok()) {
+        LOG_CTX("b6cbf", TRACE, self->loggerContext)
+            << "storage operation completed";
+        // resolve and continue
+        req.promise.setValue(std::move(result));
+        guard = self->guardedData.getLockedGuard();
+        guard->onDiskLog = std::move(req.logResult);
+        guard->onDiskMapping = std::move(req.mappingResult);
+      } else {
+        LOG_CTX("77587", ERR, self->loggerContext)
+            << "failed to commit storage operation: " << result;
+        // lock directly, and flush all the queue.
+        // If we resolve directly (which could result in a retry) we would
+        // immediately abort any retries with `precondition` failed
+        guard = self->guardedData.getLockedGuard();
+        // restore old state
+        guard->spearheadLog = guard->onDiskLog;
+        // clear queue
+        auto queue = std::move(guard->queue);
+        guard->queue.clear();
+        guard.unlock();
+        // resolve everything else
+        req.promise.setValue(std::move(result));
+        for (auto& r : queue) {
+          LOG_CTX("507fe", INFO, self->loggerContext)
+              << "aborting storage operation because of error in previous "
+                 "operation";
+          r.promise.setValue(TRI_ERROR_ARANGO_CONFLICT);
```

Comment:
Would it be prudent to choose another error code? If I saw something fail with a conflict on a follower, I'd probably panic and assume other things have gone horribly wrong. Maybe some `TRI_ERROR_REPLICATION_REPLICATED_LOG_SUBSEQUENT_FAULT` or `TRI_ERROR_REPLICATION_REPLICATED_LOG_STORAGE_TRANSACTION_ABORTED` or so...